### PR TITLE
Add csrmanager with read only

### PR DESCRIPTION
**Pull request by xiaoling-yi**
In this PR, we add the following:
* bring the stall back (when the accelerator is busy, writing 0 the start register will be stalled)
* the CsrManager with Read Only CSR using the mechanism as shown in the figure below

image

* the test for the CsrManager with Read Only CSR
* wrap commonly used test pattern as function and clear the test for the original CsrManager a bit
* some format changes caused by the `scalafmt` cmd (This mainly happens in unrelated files, please ignore them.)

### CODE REVIEW

**Comment by xiaoling-yi** [comment]
Hi @rgantonio , here we have the CsrManager with Read Only CSRs. Please check the configuration method in  `CsrManagerWithReadOnlyGenerate.scala` and `CsrManagerWithReadOnlyTestParameters.scala` under `src/test/scala/snax/csr_manager/`. When csrNumReadOnly =2, in the generated system verilog, the extra read-only CSR input port from the accelerator is  
` input  [31:0] io_read_only_csr_0,	
                io_read_only_csr_1,`
                
This module should be enough for your ALU example. If not, please let me know!

File: hw/chisel/src/main/scala/snax/csr_manager/CsrManagerWithReadOnly.scala [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    csrNumReadOnly: Int,
+    csrAddrWidth: Int
+) extends CsrManagerIO(csrNum, csrAddrWidth) {
+
+  val read_only_csr = Input(Vec(csrNumReadOnly, UInt(csrAddrWidth.W)))
+
+}
+
+class CsrManagerWithReadOnly(
+    csrNum: Int,
```
Comment (rgantonio) [inline hw/chisel/src/main/scala/snax/csr_manager/CsrManagerWithReadOnly.scala:19]: I like to re-name the `csrNum` to `csrNumReadWrite`. Reason: I remember when I was working before we had conventional signal names like `RW` - read and write, `RO` - read only, and `WO` write only. We even had the `pass_through` registers which are custom controlled by some use. What do you think? Would it be nice to re-write it as `csrNumReadWrite`?

File: hw/chisel/src/main/scala/snax/csr_manager/CsrManagerWithReadOnly.scala
Comment (rgantonio) [inline hw/chisel/src/main/scala/snax/csr_manager/CsrManagerWithReadOnly.scala:1]: Oh! For this you created 2 CSRManagers, one with read-write only and one with read-only? I got that right right?

File: hw/chisel/src/test/scala/snax/csr_manager/CsrManagerGenerate.scala [hunk 2]
```diff
@@ -4,13 +4,14 @@ import chisel3._
...
 
   emitVerilog(
-    new CsrManager(10, 32),
-    Array("--target-dir", "generated/")
-  )    
+    new CsrManager(
+      CsrManagerTestParameters.csrNum,
+      CsrManagerTestParameters.csrAddrWidth
+    ),
+    Array("--target-dir", "generated/csr_manager")
```
Comment (rgantonio) [inline hw/chisel/src/test/scala/snax/csr_manager/CsrManagerGenerate.scala:14]: For these, do we point it into the `hw/chisel/genreated/csr_manager/.` ? Or do we put it at the `../../target/snitch_cluster/generated/.` ?

File: hw/chisel/src/test/scala/snax/csr_manager/CsrManagerTest.scala
Comment (rgantonio) [inline hw/chisel/src/test/scala/snax/csr_manager/CsrManagerTest.scala:1]: Hoho such a comprehensive test. We can minimize it next time or do a separate PR for a more complex or comprehensive test 😄

File: hw/chisel/src/test/scala/snax/csr_manager/CsrManagerWithReadOnlyGenerate.scala [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+class CsrManagerWithReadOnlyGenerate extends AnyFlatSpec {
+
+  emitVerilog(
+    new CsrManagerWithReadOnly(
+      csrNum = CsrManagerWithReadOnlyTestParameters.csrNum,
+      csrNumReadOnly = CsrManagerWithReadOnlyTestParameters.csrNumReadOnly,
+      csrAddrWidth = CsrManagerWithReadOnlyTestParameters.csrAddrWidth,
+      csrModuleTagName = "test"
+    ),
+    Array("--target-dir", "generated/csr_manager")
```
Comment (rgantonio) [inline hw/chisel/src/test/scala/snax/csr_manager/CsrManagerWithReadOnlyGenerate.scala:16]: Same default place of output? `../../target/snitch_cluster/generated/.`?

File: hw/chisel/src/test/scala/snax/csr_manager/CsrManagerWithReadOnlyTest.scala [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    ).withAnnotations(Seq(WriteVcdAnnotation)) { dut =>
+      base_csr_manager_test(dut)
+
+      // ***********************************************************************
+      // test read only csr
+      // ***********************************************************************
+
+      // drive the read only csr
+      dut.io.read_only_csr(0).poke(1.U)
+      dut.io.read_only_csr(1).poke(2.U)
```
Comment (rgantonio) [inline hw/chisel/src/test/scala/snax/csr_manager/CsrManagerWithReadOnlyTest.scala:33]: hehe is there a way to do something like putting random values?

```scala
dut.io.read_only_csr(0).poke(random)
dut.io.read_only_csr(1).poke(random)
```

File: hw/chisel/src/test/scala/snax/csr_manager/CsrManagerWithReadOnlyTest.scala
Comment (rgantonio) [inline hw/chisel/src/test/scala/snax/csr_manager/CsrManagerWithReadOnlyTest.scala:1]: Can be improved here or in another PR but can we add like checking a simple case where:

1. We write to the RW registers
2. Have the RO registers set (just like in test)
3. Read from the RW to the RO continuously. I was thinking simple for loops could handle it.

File: hw/chisel/src/test/scala/snax/csr_manager/CsrManagerWithReadOnlyTestParameters.scala
Comment (rgantonio) [inline hw/chisel/src/test/scala/snax/csr_manager/CsrManagerWithReadOnlyTestParameters.scala:1]: Next PR we extend to 5 haha just to be sure 😅

File: hw/chisel/src/test/scala/snax/csr_manager/TestUtils.scala [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    dut.io.csr_config_in.req.bits.data.poke(data.U)
+    dut.io.csr_config_in.req.bits.addr.poke(addr.U)
+    dut.io.csr_config_in.req.valid.poke(1.B)
+
+    // wait for grant
+    while (dut.io.csr_config_in.req.ready.peekBoolean() == false) {
+      dut.clock.step(1)
+    }
+
+    dut.clock.step(1)
```
Comment (rgantonio) [inline hw/chisel/src/test/scala/snax/csr_manager/TestUtils.scala:27]: I think this does double clock when waiting for ready? Like:

```scala

if (dut.io.csr_config_in.req.ready.peekBoolean() == false) {
    while (dut.io.csr_config_in.req.ready.peekBoolean() == false) {
          dut.clock.step(1)
    }
} else {
    dut.clock.step(1)
}

```
The point is just to avoid double clocking haha!

And just a minor nitpick but would it be better to test for both valid and ready? I know valid was set prior (I mean duh hahaha) but like just to make sure when someone reads the code they understand you meant to check the valid-ready response.

**Review by rgantonio — COMMENTED** [review]
@xiaoling-yi NICE WORK!

I mean easy peasy addition. Please resolve comments first before moving. Let me re-review to check then I'll approve haha.

Also, the formatting you use... is it built in the terminal or in the vs code? I think we need to let everyone add that step before doing fixes in your scala code.  I noticed it's also formatting the streamer parts too.

File: hw/chisel/src/main/scala/snax/csr_manager/CsrManager.scala [hunk 6]
```diff
@@ -50,22 +50,33 @@ class CsrManager(
     with RequireAsyncReset {
   override val desiredName = csrModuleTagName + "CsrManager"
 
-  val io = IO(new CsrManagerIO(csrNum, csrAddrWidth))
+  lazy val io = IO(new CsrManagerIO(csrNum, csrAddrWidth))
+  io.suggestName("io")
```
[the remark this answers is not in the document]
Comment (xiaoling-yi) [reply hw/chisel/src/main/scala/snax/csr_manager/CsrManager.scala:54]: It is for a clear name for the generated wires.

File: hw/chisel/src/main/scala/snax/csr_manager/CsrManager.scala [hunk 7]
```diff
@@ -50,22 +50,33 @@ class CsrManager(
...
 
   // generate a vector of registers to store the csr state
   val csr = RegInit(VecInit(Seq.fill(csrNum)(0.U(32.W))))
 
   // read write and start csr command
   val read_csr = io.csr_config_in.req.fire && !io.csr_config_in.req.bits.write
   val write_csr = io.csr_config_in.req.fire && io.csr_config_in.req.bits.write
-  val start_csr = io.csr_config_in.req.bits.write &&
+  val start_csr = io.csr_config_in.req.valid && io.csr_config_in.req.bits.write &&
```
[the remark this answers is not in the document]
Comment (xiaoling-yi) [reply hw/chisel/src/main/scala/snax/csr_manager/CsrManager.scala:62]: Discussed offline. The ready signal comes directly from the accelerator.

File: hw/chisel/src/main/scala/snax/csr_manager/CsrManager.scala [hunk 8]
```diff
@@ -50,22 +50,33 @@ class CsrManager(
...
     (io.csr_config_in.req.bits.addr === (csrNum - 1).U) && io.csr_config_in.req.bits.data === 1.U
-
-  // assert the csr address is valid
-  when(io.csr_config_in.req.fire) {
-    assert(io.csr_config_in.req.bits.addr < csrNum.U, "csr address overflow!")
+  val check_acc_status =
+    io.csr_config_in.req.valid && io.csr_config_in.req.bits.write &&
+      (io.csr_config_in.req.bits.addr === (csrNum - 1).U) && io.csr_config_in.req.bits.data === 0.U
```
[the remark this answers is not in the document]
Comment (xiaoling-yi) [reply hw/chisel/src/main/scala/snax/csr_manager/CsrManager.scala:66]: Discussed offline. Idiomatic expression for the `start_csr` and `acc_check_status` so it's clear what they are.

File: hw/chisel/src/test/scala/snax/csr_manager/CsrManagerGenerate.scala (hunk 2 above)
Comment (xiaoling-yi) [reply hw/chisel/src/test/scala/snax/csr_manager/CsrManagerGenerate.scala:14]: For bug, we put it locally.

File: hw/chisel/src/main/scala/snax/csr_manager/CsrManagerWithReadOnly.scala (hunk 1 above)
Comment (xiaoling-yi) [reply hw/chisel/src/main/scala/snax/csr_manager/CsrManagerWithReadOnly.scala:19]: Changed!

File: hw/chisel/src/test/scala/snax/csr_manager/CsrManagerWithReadOnlyTest.scala (hunk 4 above)
Comment (xiaoling-yi) [reply hw/chisel/src/test/scala/snax/csr_manager/CsrManagerWithReadOnlyTest.scala:33]: Test specific data for control and random data for compute

File: hw/chisel/src/test/scala/snax/csr_manager/CsrManagerWithReadOnlyTest.scala
Comment (xiaoling-yi) [reply hw/chisel/src/test/scala/snax/csr_manager/CsrManagerWithReadOnlyTest.scala:1]: It is in the `base_csr_manager_test`

File: hw/chisel/src/test/scala/snax/csr_manager/TestUtils.scala (hunk 5 above)
Comment (xiaoling-yi) [reply hw/chisel/src/test/scala/snax/csr_manager/TestUtils.scala:27]: No, the timing is correct. Because the while loop will exit on the immediate cycle when ready is high. Therefore it does not do double clock before doing the `valid.poke`.

**Comment by xiaoling-yi** [comment]
> @xiaoling-yi NICE WORK!
> 
> I mean easy peasy addition. Please resolve comments first before moving. Let me re-review to check then I'll approve haha.
> 
> Also, the formatting you use... is it built in the terminal or in the vs code? I think we need to let everyone add that step before doing fixes in your scala code. I noticed it's also formatting the streamer parts too.

I use `scalafmt` cmd in the terminal.

File: hw/chisel/src/main/scala/snax/csr_manager/CsrManager.scala [hunk 9]
```diff
@@ -21,51 +21,83 @@ class SnaxCsrIO(csrAddrWidth: Int) extends Bundle {
 /** This class represents the input and output ports of the CsrManager module.
   * The input is connected to the SNAX CSR port. The output is connected to the
   * streamer configuration port.
-  * @param csrNum
-  *   the number of csr registers
+  * @param csrNumReadWrite
+  *   the number of read and write csr registers
```
Comment (jorendumoulin) [inline hw/chisel/src/main/scala/snax/csr_manager/CsrManager.scala:25]:
```suggestion
  *   the number of read/write csr registers. These control registers are 
  *   buffered and can only be written to by the manager core.
```

File: hw/chisel/src/main/scala/snax/csr_manager/CsrManager.scala [hunk 10]
```diff
@@ -21,51 +21,83 @@ class SnaxCsrIO(csrAddrWidth: Int) extends Bundle {
...
+  * @param csrNumReadOnly
+  *   the number of read only csr registers
```
Comment (jorendumoulin) [inline hw/chisel/src/main/scala/snax/csr_manager/CsrManager.scala:27]:
```suggestion
  *   the number of read only csr registers. These values are written to by the accelerator
```

File: hw/chisel/src/main/scala/snax/streamer/StreamerTop.scala [hunk 11]
```diff
@@ -41,34 +41,43 @@ class StreamerTop(
   override val desiredName = params.tagName + "StreamerTop"
 
   // add extra performance counter
```
Comment (jorendumoulin) [inline hw/chisel/src/main/scala/snax/streamer/StreamerTop.scala:43]: I don't think the comment is valid here anymore?
```suggestion
```

File: hw/chisel/src/main/scala/snax/streamer/StreamerTop.scala [hunk 12]
```diff
@@ -78,25 +87,8 @@ class StreamerTop(
...
-
-  // connect the streamer and csrManager output
-  // control signals
-  streamer.io.csr.valid := csr_manager.io.csr_config_out.valid
-  csr_manager.io.csr_config_out.ready := streamer.io.csr.ready
+  // connect the performance counter to the first ready only csr
+  csr_manager.io.read_only_csr(0) := performance_counter
 
   // splitting csrManager data ports to the streamer components
   // Total number of csr is temporalDim + dataMoverNum * temporalDim + spatialDim.sum + dataMoverNum + 1.
```
Comment (jorendumoulin) [inline hw/chisel/src/main/scala/snax/streamer/StreamerTop.scala:94]:
```suggestion
  //  Total number of csr is temporalDim + dataMoverNum * temporalDim + spatialDim.sum + dataMoverNum 
  // + 1 (launch address) + 1 (performance counter).
```

**Review by rgantonio — APPROVED** [review]
Noice!!!

**Comment by xiaoling-yi** [comment]
Ok, We can add the readme in the next PR!
